### DISCUSSION TITLE
perf issue (probably due to fieldState/FormState) with autosave

### DISCUSSION THREAD

**Discussion by quentinovega** [react-hook-form/react-hook-form#8535] [Q&A] [answered]
Hi,
I'm the principal maintainer of a javascript library which use react-hook-form to build forms [(@maif/react-forms)](https://github.com/MAIF/react-forms).
In the case when I activate the autosave in my lib, I experiences poor perf (maybe due to the retrieve of the formstate/fieldstate).

I extracted the logic of my code to create a [sandbox](https://codesandbox.io/s/autosave-poor-perfs-p5o7ty?file=/src/forms/AutoSave.tsx) with the problem. Is there any solution to get field errors ?
By the way, I noticed that I have always two render of my inputs after value changes, is it normal ?

**Comment by bluebill1049** [comment, marked as answer]
you can use the following subscription function if you just want to subscribe for changes.

https://react-hook-form.com/api/useform/watch

<img width="859" alt="Screen Shot 2022-06-21 at 8 33 46 am" src="https://user-images.githubusercontent.com/10513364/174684961-4e8e99f2-4f6e-4d8b-b252-cac1d3665a73.png">

```tsx
watch((data) => { data })
```

**Reply by quentinovega** [reply]
I'm not sure it will help me in this specific case, I need the state of the form or the field to get errors.
after the watch I need to call getFieldState() and I still notice poor perf when autosubmit is activated

**Reply by bluebill1049** [reply]
Could you share a simple codesandbox also worthy to check if this leads to other factors as hook form itself is very light and does not consume much rendering.

**Reply by quentinovega** [reply]
yes, that's where you see my problem: https://codesandbox.io/s/my-poor-perf-issue-with-react-hook-form-0qs1k5?file=/src/form.js. I'm rendering a very large form to reproduce my problem (it's easier than implementing an array of nested objects for example).
I'm not sure if it's due to rendering, when I run a react profiler I don't get much rendering, moreover I use a react memo to avoid this cause.
I don't if it's on cause but I noticed that when the defaultValue is just an empty object, I have better perf (not the same when I deactivate the autosubmit but...)

**Reply by bluebill1049** [reply]
```tsx
useEffect(() => {
  handleSubmit(() => {
    onSubmit(data);
  })();
}, [data]);
```

root cause above, i wouldn't call handleSubmit on each keystroke it does all the validation instead, if you just want to save temp data use `getValues` or else you should debounce the submit

**Reply by quentinovega** [reply]
Thanks @bluebill1049 , that seems to be the point of failure indeed.
I'm happy to use your project, thanks to you and everyone working on your project 🙏

**Reply by bluebill1049** [reply]
most welcome.

**Comment by Upinwv** [comment]
I’m sorry but Idk what to do I keep getting these from this thread but just can’t figure out what to do with it I think I do one moment but next moment I think I’m missing somthing??

--

**Reply by quentinovega** [reply]
@Upinwv
Basically, with my original code, the submit of the form was called every keystroke, which can be long with the validation of a large form. To fix it, I used a debounce function to only call the submit every x seconds without a keystroke.

Before my fix :
```
const Watcher = ({ control, onSubmit, handleSubmit }) => {
  const data = useWatch({ control });
  useEffect(() => {
    handleSubmit(() => {
      onSubmit(data);
    })();
  }, [data]);

  return null;
};
```

With better perf (I use a react memo to avoid rerender by my parent component and by rebound the submit without debounce):
```
const Watcher = React.memo(({ control, onSubmit, handleSubmit }) => {
  const data = useWatch({ control })

  const realSubmit = (d) => handleSubmit(() => {
    onSubmit(d);
  })()

  //debounce func from lodash
  const debouncedSubmit = useCallback(debounce(realSubmit, 250, { leading: true }), [])

  useEffect(() => {
      debouncedSubmit(data)
  }, [data])

  return null
}, () => {
  return true
})
```

PS: You can use [@maif/react-form](https://www.npmjs.com/package/@maif/react-forms) ([github repo](https://github.com/MAIF/react-forms)), now it work well 😉
